src/main/kotlin/com/openlattice/mechanic/upgrades/SyncOrgEntitySetsMissingInMeta.kt

```diff
@@ -0,0 +1,118 @@
+package com.openlattice.mechanic.upgrades
+
+import com.openlattice.data.DataGraphManager
+import com.openlattice.datastore.services.EdmManager
+import com.openlattice.datastore.services.EntitySetManager
+import com.openlattice.edm.type.PropertyType
+import com.openlattice.hazelcast.HazelcastMap
+import com.openlattice.mechanic.Toolbox
+import com.openlattice.organizations.Organization
+import com.openlattice.organizations.OrganizationMetadataEntitySetsService
+import org.apache.olingo.commons.api.edm.FullQualifiedName
+import org.slf4j.LoggerFactory
+import java.util.*
+
+class SyncOrgEntitySetsMissingInMeta(
+    private val toolbox: Toolbox,
+    private val entitySetsService: EntitySetManager,
+    private val metadataEntitySetsService: OrganizationMetadataEntitySetsService,
+    private val edmService: EdmManager,
+    private val dataService: DataGraphManager
+) : Upgrade {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(SyncOrgEntitySetsMissingInMeta::class.java)
+        private val ID_FQN = FullQualifiedName("ol.id")
+    }
+
+    override fun getSupportedVersion(): Long {
+        return Version.V2021_03_10.value
+    }
+
+    override fun upgrade(): Boolean {
+
+        val orgs = HazelcastMap.ORGANIZATIONS.getMap(toolbox.hazelcast).toMap()
+        orgs.values.forEachIndexed { index, org ->
+
+            logger.info("================================")
+            logger.info("================================")
+            logger.info("starting processing org ${org.id}")
+
+            val orgEntitySetsIds = entitySetsService.getEntitySetsForOrganization(org.id)
+            if (orgEntitySetsIds.isEmpty()) {
+                logger.info("getEntitySetsForOrganization returned empty set")
+                return@forEachIndexed
+            }
+
+            val idPT = edmService.getPropertyType(ID_FQN)
+            if (idPT == null) {
+                logger.info("ol.id PropertyType is null")
+                return@forEachIndexed
+            }
+
+            val dataSetsESID = org.organizationMetadataEntitySetIds.datasets
+            val dsEntitySet = entitySetsService.getEntitySet(dataSetsESID)
+            if (dsEntitySet == null) {
+                logger.info("\"datasets\" meta entity set is null")
+                return@forEachIndexed
+            }
+
+            val data = dataService.getEntitySetData(
+                mapOf(dataSetsESID to Optional.empty()),
+                linkedSetOf(idPT.type.toString()),
+                mapOf(dataSetsESID to mapOf(idPT.id to idPT)),
+                dsEntitySet.isLinking
+            )
+
+            if (data.entities == null) {
+                logger.info("getEntitySetData returned null entities")
+                return@forEachIndexed
+            }
+
+            val ids = data.entities
+                .mapNotNull { entity -> entity }
```

Comment:
super not important, but for future reference, you can replace this with `.filterNotNull()`